bundle/edu.gemini.epics.acm/src/main/java/edu/gemini/epics/acm/CaCommandSenderImpl.java

```diff
@@ -0,0 +1,295 @@
+            String description) throws CaException {
+        CaParameterImpl<Float> param = floatParameters.get(name);
+        if (param == null) {
+            if (alreadyExist(name)) {
+                throw new CaException(
+                        "Parameter already exists with a different type.");
+            } else {
+                param = CaParameterImpl.createFloatParameter(name, channel,
+                        description, epicsWriter);
+                floatParameters.put(name, param);
+            }
+        } else {
+            if (!channel.equals(param.channel())) {
+                throw new CaException(
+                        "Parameter already exists for a different channel.");
+            }
+        }
+        return param;
+    }
+
+    @Override
+    public CaParameter<String> addString(String name, String channel)
+            throws CaException {
+        return addString(name, channel, null);
+    }
+
+    @Override
+    public CaParameter<String> addString(String name, String channel,
+            String description) throws CaException {
+        CaParameterImpl<String> param = stringParameters.get(name);
+        if (param == null) {
+            if (alreadyExist(name)) {
+                throw new CaException(
+                        "Parameter already exists with a different type.");
+            } else {
+                param = CaParameterImpl.createStringParameter(name, channel,
+                        description, epicsWriter);
+                stringParameters.put(name, param);
+            }
+        } else {
+            if (!channel.equals(param.channel())) {
+                throw new CaException(
+                        "Parameter already exists for a different channel.");
+            }
+        }
+        return param;
+    }
+
+    @Override
+    public CaParameter<Integer> getInteger(String name) {
+        return integerParameters.get(name);
+    }
+
+    @Override
+    public CaParameter<Double> getDouble(String name) {
+        return doubleParameters.get(name);
+    }
+
+    @Override
+    public CaParameter<Float> getFloat(String name) {
+        return floatParameters.get(name);
+    }
+
+    @Override
+    public CaParameter<String> getString(String name) {
+        return stringParameters.get(name);
+    }
+
+    @Override
+    public void remove(String name) {
+        doubleParameters.remove(name);
+        floatParameters.remove(name);
+        integerParameters.remove(name);
+        stringParameters.remove(name);
+    }
+
+    private boolean alreadyExist(String name) {
+        if (doubleParameters.containsKey(name)
+                || floatParameters.containsKey(name)
+                || integerParameters.containsKey(name)
+                || stringParameters.containsKey(name)) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public void mark() throws TimeoutException {
+        if (dirChannel != null) {
```

Comment:
There are several instances of null checks that could be removed if the null were forbidden right at the constructor and final, unless of course null is a valid value